src/resolve/resolved-dns-stub.c

```diff
@@ -632,50 +638,67 @@ static int manager_dns_stub_udp_fd_extra(Manager *m, DnsStubListenerExtra *l) {
                         .in6.sin6_addr = l->address.in6,
                 };
 
-        fd = socket(l->family, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        fd = socket(l->family, type | SOCK_CLOEXEC | SOCK_NONBLOCK, 0);
         if (fd < 0) {
                 r = -errno;
                 goto fail;
         }
 
-        if (l->family == AF_INET) {
-                r = setsockopt_int(fd, IPPROTO_IP, IP_FREEBIND, true);
-                if (r < 0)
-                        goto fail;
-        }
-
         r = set_dns_stub_common_socket_options(fd, l->family);
         if (r < 0)
                 goto fail;
 
+        /* Do not set IP_TTL for extra DNS stub listners, as the address may not be local and in that case
+         * people may want ttl > 1. */
+
+        if (l->family == AF_INET)
+                r = setsockopt_int(fd, IPPROTO_IP, IP_FREEBIND, true);
+        else
+                r = setsockopt_int(fd, IPPROTO_IPV6, IPV6_FREEBIND, true);
+        if (r < 0)
+                goto fail;
```

Comment:
this would be much prettier if based on #17005's socket_set_freebind() helper!

Reply:
Interesting. One person touches a piece of code and then it cascades into a series of lengthy prs. I guess this code was ripe for some cleanups.